### DISCUSSION TITLE
Investigate type reference capabilities for validation (constraints)

### DISCUSSION THREAD

**Discussion by lizzcondrey** [DevOps-MBSE/AaC#701] [Ideas]
As we evaluate some engine and core functionality improvements, type references has come up. To determine capability and implementation paths, we are wanting to investigate incorporating type referencing within the validation (constraint) system of AaC to assist with reducing the complexity of the current validation (constraint) system. Incorporating type referencing should make readability and maintainability easier for the validation (constraint) system by introducing a common coding concept.

**Comment by jondavid-black** [comment]
My current thoughts on the topic are to provide a primitive type ```reference``` that is defined with a parameter via parenthesis.  As an example, if a user want the field they're defining to reference a requirement id they would declare the type of that field as ```reference(req.id)```.  The primitive constraint for this would look through definitions in the LanguageContext that have a ```root``` of req with a field named ```id``` that contains the value in the instance of the field.  If found, the reference is valid, otherwise there should be a constraint violation.  A separate constraint we'll need is to ensure the parameter provided to the reference is valid in the language definition.

Do you believe this approach to defining references meets the needs you describe?

**Comment by lizzcondrey** [comment]
This is currently being further evaluated and experimentally implemented within the AaC 4.0 work going on. So more thorough review and consideration will be able to be conducted once that experiment implementation is in place to determine any remaining holes.
